libs/ml/include/ml/optimisation/lazy_adam_optimiser.hpp

```diff
@@ -0,0 +1,371 @@
+
+  template <typename X, typename D>
+  friend struct serializers::MapSerializer;
+
+  OptimiserType OptimiserCode() override
+  {
+    return OptimiserType::ADAM;
+  }
+
+private:
+  std::vector<TensorType> cache_;
+  std::vector<TensorType> momentum_;
+  std::vector<TensorType> mt_;
+  std::vector<TensorType> vt_;
+
+  DataType beta1_;
+  DataType beta2_;
+  DataType beta1_t_;
+  DataType beta2_t_;
+
+  // ApplyGradientSparse if number_of_rows_to_update*sparsity_threshold_ <= total_rows
+  SizeType sparsity_threshold_ = 4;
+
+  DataType epsilon_;
+
+  void ResetCache();
+  void Init();
+  void ApplyLogic(SizeType batch_size, TensorType &gradient_tensor, TensorType &momentum_tensor,
+                  TensorType &mt_tensor, TensorType &v_tensor, TensorType &cache_tensor,
+                  TensorType const &refs_tensor);
+};
+
+template <class T>
+void LazyAdamOptimiser<T>::Init()
+{
+  for (auto &train : this->graph_trainables_)
+  {
+    this->cache_.emplace_back(TensorType(train->GetWeights().shape()));
+    this->momentum_.emplace_back(TensorType(train->GetWeights().shape()));
+    this->mt_.emplace_back(TensorType(train->GetWeights().shape()));
+    this->vt_.emplace_back(TensorType(train->GetWeights().shape()));
+  }
+  ResetCache();
+}
+
+template <class T>
+LazyAdamOptimiser<T>::LazyAdamOptimiser(std::shared_ptr<Graph<T>>       graph,
+                                        std::vector<std::string> const &input_node_names,
+                                        std::string const &             label_node_name,
+                                        std::string const &             output_node_name,
+                                        DataType const &learning_rate, DataType const &beta1,
+
+                                        DataType const &beta2, SizeType sparsity_threshold,
+                                        DataType const &epsilon)
+  : Optimiser<T>(graph, input_node_names, label_node_name, output_node_name, learning_rate)
+  , beta1_(beta1)
+  , beta2_(beta2)
+  , beta1_t_(beta1)
+  , beta2_t_(beta2)
+  , sparsity_threshold_(sparsity_threshold)
+  , epsilon_(epsilon)
+{
+  Init();
+}
+
+template <class T>
+LazyAdamOptimiser<T>::LazyAdamOptimiser(
+    std::shared_ptr<Graph<T>> graph, std::vector<std::string> const &input_node_names,
+    std::string const &label_node_name, std::string const &output_node_name,
+    fetch::ml::optimisers::LearningRateParam<DataType> const &learning_rate_param,
+    DataType const &beta1, DataType const &beta2, SizeType sparsity_threshold,
+    DataType const &epsilon)
+  : Optimiser<T>(graph, input_node_names, label_node_name, output_node_name, learning_rate_param)
+  , beta1_(beta1)
+  , beta2_(beta2)
+  , beta1_t_(beta1)
+  , beta2_t_(beta2)
+  , sparsity_threshold_(sparsity_threshold)
+  , epsilon_(epsilon)
+{
+  Init();
+}
+
+// private
+
+template <class T>
+void LazyAdamOptimiser<T>::ApplyLogic(SizeType batch_size, TensorType &gradient_tensor,
```

Comment:
apply logic means what exactly?
Please add a function comment describing the responsibility of this function

Reply:
e.g. updates momenta?